Learn HTML (modules 12-20) (#9576)
GoogleChrome/web.dev: src/site/content/en/learn/html/images/index.md

not a description of the image. For example, the magnifying glass icon's alt attribute is `search`. The icon which looks
like a house has `home` as the alt text. The 5-inch floppy disc icon's description is `save`. If there are two icons of Fluffy used
to indicate best practices and anti-patterns, the smiling dog in a green beret could have `alt="good"` set, while the snarling dog
in a red beret might read `alt="bad"`. That said, only use standard iconography, and if you use non-standard icons such as
the good and bad Fluffy, include a legend and ensure that the icons are not the only ways of deciphering the meaning of your UI elements,

If the image is a screenshot or a graph, write what is learned from the image rather than describing the appearance.
While an image can definitely be worth a thousand words, the description should concisely convey everything that is learned.

Omit information the user already knows from the context and is otherwise informed about in the content. For example,
if you're on a tutorial page about changing browser settings and the page is about clicking icons in the browser chrome, the URL

> of the page in the screen capture isn't important. Limit the alt to the topic at hand: how to change settings. The alt might be

Reviewer: ```suggestion
of the page in the screen capture isn't important. Limit the `alt` to the topic at hand: how to change settings. The `alt` might be
```

as the user doesn't need to know it's a screenshot and doesn't need to know where the techwriter was surfing when they wrote
the instructions. The description of the image is based on the context of why the image was included in the first place.

If the screen capture shows how to find the browser version number by going to `chrome://version/`, include the URL in the
content of the page as instructions, and provide an empty string as the alt attribute as the image provides no information
that is not in the surrounding text.

If the image provides no additional information or is purely decorative, the attribute should still be there, just as an empty string.

```html
<img src="svg/magazine.svg" alt="" role="none" />
```

MachineLearningWorkshop.com has seven foreground images, therefore seven images with alt attributes: an easter egg light switch,
a manual icon, two biographical photos of Hal and Eve, and three avatars of a blender, a vacuum cleaner, and a toaster. The
foreground image that looks like a magazine is the only one that is purely decorative. The page also has
two background images; these are also decorative and, as they are added with CSS, are inaccessible.

> The magazine, being purely decorative, has an empty `alt` attribute, and a `role` of `none` as the image is a purely

Reviewer: ```suggestion
The magazine, being purely decorative, has an empty `alt` attribute, and a [`role` of `none`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/presentation_role) as the image is a purely
```

## Responsive images

There are a myriad of viewport sizes. There are also different screen resolutions. You don't want to waste a mobile user's
bandwidth by serving them an image wide enough for a large screen monitor, but you might want to serve higher resolution
images for tiny devices that have four times the normal screen resolution. There are a few ways to serve different images
based on the viewport size and the screen resolution.

### `<img> srcset` attribute

The [`srcset`](/learn/design/responsive-images/#responsive-images-with-srcset) attribute enables suggesting multiple image files,
with the browser selecting which image to request based on multiple media queries including viewport size and screen resolution.

There can be a single `srcset` attribute per `<img>` element, but that `srcset` can link to multiple images. The `srcset`
attribute accepts a list of comma-separated values, each containing the URL of the asset followed by a space followed by
descriptors for that image option. If a width descriptor is used, you must also include the `sizes` attribute with a media

> query or source size for each `srcset` option other than the last one. The Learn section covering [responsive images with `srcset`](/learn/design/responsive-images/#responsive-images-with-srcset)
> is worth reading.

Reviewer: ```suggestion
query or source size for each `srcset` option other than the last one. The Learn sections covering [responsive images with `srcset`](/learn/design/responsive-images/#responsive-images-with-srcset) and [descriptive syntaxes](/learn/images/descriptive/) are worth reading.
is worth reading.
```

The `srcset` image will take precedence over the `src` image if there is a match.

### `<picture>` and `<source>`

Another method for providing multiple resources and allowing the browser to render the most appropriate asset is with the
[`<picture>`](https://developer.mozilla.org/docs/Web/HTML/Element/picture) element. The `<picture>` element is a container
element for multiple image options listed in an unlimited number of [`<source>`](https://developer.mozilla.org/docs/Web/HTML/Element/source)
elements and a single required [`<img>`](https://developer.mozilla.org/docs/Web/HTML/Element/img) element.

The [`<source>`](https://developer.mozilla.org/docs/Web/HTML/Element/source) attributes include `srcset`, `sizes`, `media`, `width`, and `height`.
The `srcset` attribute is common to `img`, `source`, and `link`, but is generally implemented slightly differently on source
as media queries can be listed in the `<srcset>`'s media attribute instead. `<source>` also supports image formats defined in the `type` attribute.

The browser will consider each child `<source>` element and choose the best match among them. If no matches are found, the URL

> of the `<img>` element's [`src`](https://developer.mozilla.org/docs/Web/HTML/Element/img#attr-src) attribute is selected. The accessible name comes from the `alt.attribute` of the nested `<img>`.\

Reviewer: ```suggestion
of the `<img>` element's [`src`](https://developer.mozilla.org/docs/Web/HTML/Element/img#attr-src) attribute is selected. The accessible name comes from the `alt` attribute of the nested `<img>`.
```

> The Learn section covering the [`<picture>`](/learn/design/picture-element/) element is also worth a read.

Reviewer: ```suggestion
The Learn sections covering the [`<picture>`](/learn/design/picture-element/) element and [prescriptive syntaxes](/learn/images/prescriptive/) are also worth a read.
```

`height` and `width` values will be overridden with CSS. The purpose of including these attributes is to reserve the space at
the right aspect ratio, improving performance by reducing layout shift. The page will still take approximately the same amount
of time to load, but the UI won't jump when the image is painted to the screen.

## Other image features

The `<img>` element also supports the `crossorigin`, `decoding`, `referrerpolicy`, and, in Blink-based browsers,
`fetchpriority` attributes. Rarely used, if the image is part of a server-side map, include the `ismap` boolean attribute
and nest the `<img>` in a link for users without pointing devices.

The `ismap` attribute isn't necessary, or even supported, on the `<input type="image" name="imageSubmitName">` as the `x` and `y`
coordinates of the click location are sent during form submission, appending the values to the input name, if any. For example,
something like `&imageSubmitName.x=169&imageSubmitName.y=66` will be submitted with the form when the user clicks the image,
submitting it. If the image doesn't have a `name` attribute, the x and y are sent: `&x=169&y=66`.

> 
> ## Check your understanding

Reviewer: ```suggestion
```
